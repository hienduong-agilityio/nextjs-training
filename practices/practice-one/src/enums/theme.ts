export enum BUTTON_VARIANTS {
  'SOLID' = 'solid',
  'SHADOW' = 'shadow',
  'LIGHT' = 'light',
}

export enum BUTTON_COLORS {
  'DEFAULT' = 'default',
  'PRIMARY' = 'primary',
  'SECONDARY' = 'secondary',
  'SUCCESS' = 'success',
  'WARNING' = 'warning',
  'DANGER' = 'danger',
}
