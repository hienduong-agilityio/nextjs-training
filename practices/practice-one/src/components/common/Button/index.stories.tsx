import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

// Components
import Button from '.';

// Enums
import { BUTTON_VARIANTS, BUTTON_COLORS } from '@/enums';

const meta = {
  title: 'Components/Common/Button',
  component: Button,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    disabled: {
      control: 'boolean',
      type: 'boolean',
      description: 'Disables the button when set to `true`.',
    },
    variant: {
      control: 'inline-radio',
      options: [
        BUTTON_VARIANTS.SOLID,
        BUTTON_VARIANTS.SHADOW,
        BUTTON_VARIANTS.LIGHT,
      ],
      description: 'Defines the visual style of the button.',
    },
    color: {
      control: 'inline-radio',
      options: [
        BUTTON_COLORS.DEFAULT,
        BUTTON_COLORS.PRIMARY,
        BUTTON_COLORS.SECONDARY,
        BUTTON_COLORS.SUCCESS,
        BUTTON_COLORS.WARNING,
        BUTTON_COLORS.DANGER,
      ],
      description: 'Defines the color of the button.',
    },
    onClick: {
      action: 'clicked',
      description: 'Callback function triggered when the button is clicked.',
    },
    children: {
      control: 'text',
      description: 'Content displayed inside the button.',
    },
    startIcon: {
      control: 'text',
      description: 'Optional icon displayed at the start of the button.',
    },
    type: {
      control: 'text',
      type: 'string',
      description: 'Select button type.',
    },
    endIcon: {
      control: 'text',
      description: 'Optional icon displayed at the end of the button.',
    },
    customClass: {
      control: 'text',
      description: 'Additional custom CSS classes for the button.',
    },
  },
  args: {
    onClick: fn(),
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: 'Default Button',
    variant: BUTTON_VARIANTS.SOLID,
    color: BUTTON_COLORS.DEFAULT,
  },
  parameters: {
    docs: {
      description: {
        story:
          'This is the default button with solid variant and default color.',
      },
    },
  },
};

export const Icon: Story = {
  args: {
    children: '🔍',
    variant: BUTTON_VARIANTS.SOLID,
    customClass: 'p-2',
  },
  render: () => (
    <div className="flex gap-5">
      <Button startIcon="🔍" variant={BUTTON_VARIANTS.SOLID} customClass="p-2">
        Search
      </Button>
      <Button endIcon="🚀" variant={BUTTON_VARIANTS.SOLID} customClass="p-2">
        Launch
      </Button>
      <Button
        startIcon="✔️"
        endIcon="➡️"
        variant={BUTTON_VARIANTS.SOLID}
        customClass="p-2"
      >
        Confirm
      </Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'An icon button with solid style.',
      },
    },
  },
};

export const Sizes: Story = {
  args: {
    children: 'Large Button',
    customClass: 'text-lg p-4',
    variant: BUTTON_VARIANTS.SOLID,
    color: BUTTON_COLORS.DANGER,
  },
  render: () => (
    <div className="flex flex-col gap-3">
      <Button customClass="text-sm p-2">Small Button</Button>
      <Button customClass="text-base p-3">Medium Button</Button>
      <Button customClass="text-lg p-4">Large Button</Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'A button with custom size styles and danger color.',
      },
    },
  },
};

export const SolidVariantColors: Story = {
  args: {
    children: '',
    variant: BUTTON_VARIANTS.SOLID,
  },

  render: () => (
    <div className="flex gap-5">
      <Button color={BUTTON_COLORS.PRIMARY}>Primary</Button>
      <Button color={BUTTON_COLORS.SECONDARY}>Secondary</Button>
      <Button color={BUTTON_COLORS.SUCCESS}>Success</Button>
      <Button color={BUTTON_COLORS.WARNING}>Warning</Button>
      <Button color={BUTTON_COLORS.DANGER}>Danger</Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Displays buttons with solid variant across different colors.',
      },
    },
  },
};

export const ShadowVariantColors: Story = {
  args: {
    children: '',
    variant: BUTTON_VARIANTS.SHADOW,
  },
  render: (args) => (
    <div className="flex gap-5">
      <Button {...args} color={BUTTON_COLORS.PRIMARY}>
        Primary
      </Button>
      <Button {...args} color={BUTTON_COLORS.SECONDARY}>
        Secondary
      </Button>
      <Button {...args} color={BUTTON_COLORS.SUCCESS}>
        Success
      </Button>
      <Button {...args} color={BUTTON_COLORS.WARNING}>
        Warning
      </Button>
      <Button {...args} color={BUTTON_COLORS.DANGER}>
        Danger
      </Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Displays buttons with shadow variant across different colors.',
      },
    },
  },
};

export const LightVariantColors: Story = {
  args: {
    children: '',
    variant: BUTTON_VARIANTS.LIGHT,
  },
  render: (args) => (
    <div className="flex gap-5">
      <Button {...args} color={BUTTON_COLORS.PRIMARY}>
        Primary
      </Button>
      <Button {...args} color={BUTTON_COLORS.SECONDARY}>
        Secondary
      </Button>
      <Button {...args} color={BUTTON_COLORS.SUCCESS}>
        Success
      </Button>
      <Button {...args} color={BUTTON_COLORS.WARNING}>
        Warning
      </Button>
      <Button {...args} color={BUTTON_COLORS.DANGER}>
        Danger
      </Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Displays buttons with light variant across different colors.',
      },
    },
  },
};
