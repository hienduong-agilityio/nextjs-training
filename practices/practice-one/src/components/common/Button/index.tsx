// Libraries
import { memo } from 'react';
import { twMerge } from 'tailwind-merge';
import classNames from 'classnames';

// Types
import type { ButtonHTMLAttributes, ReactNode } from 'react';

// Enums
import { BUTTON_VARIANTS, BUTTON_COLORS } from '@/enums';

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  children: ReactNode;
  startIcon?: ReactNode;
  endIcon?: ReactNode;
  variant?: BUTTON_VARIANTS;
  color?: BUTTON_COLORS;
  customClass?: string;
}

const baseClass =
  'flex justify-center gap-2 rounded p-2 disabled:cursor-not-allowed disabled:opacity-50';

const variantClasses: Record<BUTTON_VARIANTS, string> = {
  [BUTTON_VARIANTS.SOLID]: 'bg-opacity-100 shadow-md',
  [BUTTON_VARIANTS.SHADOW]: 'shadow-lg bg-white hover:shadow-xl',
  [BUTTON_VARIANTS.LIGHT]: 'bg-gray-100 bg-opacity-50 hover:bg-opacity-70',
};

const colorClasses: Record<BUTTON_COLORS, string> = {
  [BUTTON_COLORS.DEFAULT]: 'bg-gray-200 text-black hover:bg-gray-300',
  [BUTTON_COLORS.PRIMARY]: 'bg-primary-300 text-white hover:bg-primary-400',
  [BUTTON_COLORS.SECONDARY]: 'bg-gray-400 text-white hover:bg-gray-500',
  [BUTTON_COLORS.SUCCESS]: 'bg-green-600 text-white hover:bg-green-700',
  [BUTTON_COLORS.WARNING]: 'bg-amber-400 text-black hover:bg-amber-500',
  [BUTTON_COLORS.DANGER]: 'bg-rose-500 text-white hover:bg-rose-600',
};

/**
 * Button component
 *
 * @returns {JSX.Element} - The rendered button element.
 */
const Button = ({
  children,
  startIcon = '',
  endIcon = '',
  type = 'button',
  variant = BUTTON_VARIANTS.SOLID,
  color = BUTTON_COLORS.DEFAULT,
  customClass = '',
  ...rest
}: IButtonProps): JSX.Element => {
  const combinedClasses = twMerge(
    classNames(baseClass, variantClasses[variant], colorClasses[color]),
    customClass,
  );

  return (
    <button type={type} className={combinedClasses} {...rest}>
      {startIcon && <span>{startIcon}</span>}
      {children}
      {endIcon && <span>{endIcon}</span>}
    </button>
  );
};

export default memo(Button);
