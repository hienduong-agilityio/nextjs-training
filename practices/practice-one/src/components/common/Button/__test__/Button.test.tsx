import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

// Enums
import { BUTTON_COLORS, BUTTON_VARIANTS } from '@/enums';

// Components
import Button from '..';

describe('Button Component', () => {
  let renderButton: (
    props?: Partial<React.ComponentProps<typeof Button>>,
  ) => ReturnType<typeof render>;

  beforeEach(() => {
    renderButton = (
      props: Partial<React.ComponentProps<typeof Button>> = {},
    ) => {
      return render(<Button {...props}>Test Button</Button>);
    };
  });

  it('renders with default props and matches snapshot', () => {
    const { container } = renderButton();
    const button = screen.getByRole('button', { name: /test button/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveClass('bg-gray-200 text-black');
    expect(container).toMatchSnapshot();
  });

  it('renders with different BUTTON_VARIANTS', () => {
    const { rerender } = render(
      <Button variant={BUTTON_VARIANTS.SOLID}>Test Button</Button>,
    );
    expect(screen.getByRole('button', { name: /test button/i })).toHaveClass(
      'shadow-md',
    );

    rerender(<Button variant={BUTTON_VARIANTS.SHADOW}>Test Button</Button>);
    expect(screen.getByRole('button', { name: /test button/i })).toHaveClass(
      'shadow-lg',
    );

    rerender(<Button variant={BUTTON_VARIANTS.LIGHT}>Test Button</Button>);
    expect(screen.getByRole('button', { name: /test button/i })).toHaveClass(
      'bg-gray-200 bg-opacity-50',
    );
  });

  it('renders with different colors', () => {
    const { rerender } = render(
      <Button color={BUTTON_COLORS.PRIMARY}>Test Button</Button>,
    );
    expect(screen.getByRole('button', { name: /test button/i })).toHaveClass(
      'bg-primary-300',
    );

    rerender(<Button color={BUTTON_COLORS.SUCCESS}>Test Button</Button>);
    expect(screen.getByRole('button', { name: /test button/i })).toHaveClass(
      'bg-green-600',
    );

    rerender(<Button color={BUTTON_COLORS.DANGER}>Test Button</Button>);
    expect(screen.getByRole('button', { name: /test button/i })).toHaveClass(
      'bg-rose-500',
    );
  });

  it('handles onClick event', () => {
    const handleClick = jest.fn();
    renderButton({ onClick: handleClick });

    const button = screen.getByRole('button', { name: /test button/i });
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('renders with start and end icons', () => {
    renderButton({ startIcon: <span>Start</span>, endIcon: <span>End</span> });

    expect(screen.getByText(/start/i)).toBeInTheDocument();
    expect(screen.getByText(/end/i)).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: /test button/i }),
    ).toBeInTheDocument();
  });

  it('applies custom classes', () => {
    renderButton({ customClass: 'custom-class' });

    const button = screen.getByRole('button', { name: /test button/i });
    expect(button).toHaveClass('custom-class');
  });

  it('renders disabled button', () => {
    renderButton({ disabled: true });

    const button = screen.getByRole('button', { name: /test button/i });
    expect(button).toBeDisabled();
    expect(button).toHaveClass('disabled:bg-gray-100');
  });

  it('supports different button types', () => {
    renderButton({ type: 'submit' });

    const button = screen.getByRole('button', { name: /test button/i });
    expect(button).toHaveAttribute('type', 'submit');
  });
});
