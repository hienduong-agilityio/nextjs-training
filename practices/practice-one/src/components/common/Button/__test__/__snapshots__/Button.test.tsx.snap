// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders with default props and matches snapshot 1`] = `
<div>
  <button
    class="flex items-center gap-2 rounded p-2 disabled:cursor-not-allowed bg-opacity-100 shadow-md bg-gray-200 text-black hover:bg-gray-300 disabled:bg-gray-100"
    type="button"
  >
    Test Button
  </button>
</div>
`;
